modules/mono/mono_gd/gd_hostfxr.h

```diff
@@ -0,0 +1,98 @@
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef GD_HOSTFXR_H
+#define GD_HOSTFXR_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+
+#if defined(_WIN32)
+#define CORECLR_DELEGATE_CALLTYPE __stdcall
+#define HOSTFXR_CALLTYPE __cdecl
+#ifdef _WCHAR_T_DEFINED
+typedef wchar_t char_t;
+#else
+typedef unsigned short char_t;
+#endif
+#else
+#define CORECLR_DELEGATE_CALLTYPE
+#define HOSTFXR_CALLTYPE
+typedef char char_t;
+#endif
+
+typedef void *hostfxr_handle;
+
+enum hostfxr_delegate_type {
+	hdt_com_activation,
+	hdt_load_in_memory_assembly,
+	hdt_winrt_activation,
+	hdt_com_register,
+	hdt_com_unregister,
+	hdt_load_assembly_and_get_function_pointer,
+	hdt_get_function_pointer,
+};
+
+typedef int32_t(HOSTFXR_CALLTYPE *hostfxr_initialize_for_dotnet_command_line_fn)(
+		int argc,
+		const char_t **argv,
+		const struct hostfxr_initialize_parameters *parameters,
+		/*out*/ hostfxr_handle *host_context_handle);
+
+typedef int32_t(HOSTFXR_CALLTYPE *hostfxr_initialize_for_runtime_config_fn)(
+		const char_t *runtime_config_path,
+		const struct hostfxr_initialize_parameters *parameters,
+		/*out*/ hostfxr_handle *host_context_handle);
+
+typedef int32_t(HOSTFXR_CALLTYPE *hostfxr_get_runtime_delegate_fn)(
+		const hostfxr_handle host_context_handle,
+		enum hostfxr_delegate_type type,
+		/*out*/ void **delegate);
+
+typedef int32_t(HOSTFXR_CALLTYPE *hostfxr_close_fn)(const hostfxr_handle host_context_handle);
+
+#define UNMANAGEDCALLERSONLY_METHOD ((const char_t *)-1)
+
+typedef int(CORECLR_DELEGATE_CALLTYPE *load_assembly_and_get_function_pointer_fn)(
+		const char_t *assembly_path,
+		const char_t *type_name,
+		const char_t *method_name,
+		const char_t *delegate_type_name,
+		void *reserved,
+		/*out*/ void **delegate);
```

Comment:
Here I cherry-picked only the declarations we need. Alternatively, we could add these two headers to our repo:
https://github.com/dotnet/runtime/blob/main/src/native/corehost/coreclr_delegates.h
https://github.com/dotnet/runtime/blob/main/src/native/corehost/hostfxr.h

Reply:
Seems fine. Would be good to add a link to the reference implementations our code is based upon, and if there's significant copy-paste we should probably add the .NET license header (it's also MIT licensed, but it's "Copyright (c) .NET Foundation and Contributors" and not Godot contributors).

Reply:
Are you talking about this header or the resolver implementation?

This header is a copy-paste of the declarations we need from the two headers I listed before, but I'm thinking we may as well add these two full headers if this can be a problem.

The resolver implementation heavily mimics `libnethost` and for that I had to use their implementation as reference (I also kept the same names to make comparing easier in case they change install locations in the future). Whether that would be deemed copy-paste, I don't know. Here are links for the code I used as reference:

- [`try_get_path_from_dotnet_root`](https://github.com/dotnet/runtime/blob/20e5237104ddd178f12bc501352e45a5e91e5048/src/native/corehost/fxr_resolver.cpp#L146-L158)
- [`try_get_path`](https://github.com/dotnet/runtime/blob/20e5237104ddd178f12bc501352e45a5e91e5048/src/native/corehost/fxr_resolver.cpp#L55-L134)
- [`get_latest_fxr`](https://github.com/dotnet/runtime/blob/20e5237104ddd178f12bc501352e45a5e91e5048/src/native/corehost/fxr_resolver.cpp#L12-L52)
- [`get_dotnet_root_from_env`](https://github.com/dotnet/runtime/blob/56095347ca1bacbc7a95e468f03f02ea808b9d9b/src/native/corehost/hostmisc/utils.cpp#L402-L426)
- `get_dotnet_self_registered_dir` ([Windows](https://github.com/dotnet/runtime/blob/20e5237104ddd178f12bc501352e45a5e91e5048/src/native/corehost/hostmisc/pal.windows.cpp#L394-L462) / [Unix](https://github.com/dotnet/runtime/blob/9d6396deb02161f5ee47af72ccac52c2e1bae458/src/native/corehost/hostmisc/pal.unix.cpp#L491-L542))
  - [`get_install_location_from_file`](https://github.com/dotnet/runtime/blob/9d6396deb02161f5ee47af72ccac52c2e1bae458/src/native/corehost/hostmisc/pal.unix.cpp#L455-L489)
- `get_default_installation_dir` ([Windows](https://github.com/dotnet/runtime/blob/20e5237104ddd178f12bc501352e45a5e91e5048/src/native/corehost/hostmisc/pal.windows.cpp#L289-L350) / [Unix](https://github.com/dotnet/runtime/blob/9d6396deb02161f5ee47af72ccac52c2e1bae458/src/native/corehost/hostmisc/pal.unix.cpp#L558-L598))

If we are going to add the .NET license header, should it be alone or alongside ours? 

Their versioning follows SemVer according to [a comment](https://github.com/dotnet/runtime/blob/20e5237104ddd178f12bc501352e45a5e91e5048/src/native/corehost/fxr/fx_ver.h#L9), so I improvised a SemVer implementation. It uses both the precedence logic and the regex listed in the official SemVer site.

Reply:
You can add the .NET license header where relevant after Godot's own, like we do e.g. for `gjk_epa.cpp` which is modified for Godot but originally derived from Bullet: https://github.com/godotengine/godot/blob/master/servers/physics_3d/gjk_epa.cpp